src/h5web/providers/hsds/utils.ts

```diff
@@ -26,3 +35,55 @@ export function assertHsdsDataset(
     throw new Error('Expected entity to be HSDS dataset');
   }
 }
+
+export function processHsdsBaseType(hsdsBaseType: HsdsBaseType): HDF5BaseType {
+  if (hsdsBaseType.class === HDF5TypeClass.String) {
+    return hsdsBaseType;
+  }
+
+  const { class: hsdsClass, base } = hsdsBaseType;
+
+  const regex = /H5T_[A-z]+_([A-z])(\d+)(B|L)E/u;
+  const matches = regex.exec(base);
+
+  if (!matches) {
+    throw new Error(`Unrecognized base ${base}`);
+  }
+
+  const [, sign, length, endian] = matches;
+
+  return {
+    class: hsdsClass,
+    endianness: endian === 'L' ? Endianness.Little : Endianness.Big,
+    length: Number.parseInt(length, 10),
+    unsigned: sign === 'U' || undefined,
+  };
+}
+
+export function processHsdsType(hsdsType: HsdsType): HDF5Type {
+  if (isString(hsdsType)) {
+    return hsdsType;
+  }
+
+  if (
+    hsdsType.class === HDF5TypeClass.Array ||
+    hsdsType.class === HDF5TypeClass.VLen
+  ) {
+    return {
+      ...hsdsType,
+      base: processHsdsBaseType(hsdsType.base),
+    };
+  }
+
+  if (hsdsType.class === HDF5TypeClass.Compound) {
+    return {
+      ...hsdsType,
+      fields: hsdsType.fields.map((v) => ({
+        name: v.name,
+        type: processHsdsType(v.type),
+      })),
+    };
+  }
```

Comment:
These `if`s may be better written as a `switch`. Also, with `...hsdsType`, you keep the `base` property in the returned objects, which is unnecessary -- you could destructure `hsdsType` to get the `class` and specify it explicitly in each returned object.